src/clients/dotnet/TigerBeetle/NativeClient.cs

```diff
@@ -98,60 +94,35 @@ private static NativeClient CallInit(InitFunction initFunction, UInt128Extension
     {
         using (var memoryHandler = batch.Pin())
         {
-            var asyncRequest = new AsyncRequest<TResult, TBody>(this, operation);
+            var asyncRequest = new AsyncRequest<TResult, TBody>(operation);
 
             unsafe
             {
-                asyncRequest.Submit(memoryHandler.Pointer, batch.Length);
+                asyncRequest.Submit(this, memoryHandler.Pointer, batch.Length);
             }
 
             return await asyncRequest.Wait().ConfigureAwait(continueOnCapturedContext: false);
         }
     }
 
-    public void ReleasePacket(Packet packet)
-    {
-        unsafe
-        {
-            // It is unexpected for the client to be disposed here
-            // Since we wait for all acquired packets to be submitted and returned before disposing.
-            AssertTrue(client != IntPtr.Zero, "Client is closed");
-            AssertTrue(packet.Pointer != null, "Null packet pointer");
-            tb_client_release_packet(client, packet.Pointer);
-        }
-    }
-
-    public void Submit(Packet packet)
-    {
-        unsafe
-        {
-            // It is unexpected for the client to be disposed here
-            // Since we wait for all acquired packets to be submitted and returned before disposing.
-            AssertTrue(client != IntPtr.Zero, "Client is closed");
-            tb_client_submit(client, packet.Pointer);
-        }
-    }
-
-    public Packet AcquirePacket()
+    public unsafe void Submit(TBPacket* packet)
     {
         unsafe
         {
-            if (client == IntPtr.Zero) throw new ObjectDisposedException("Client is closed");
-
-            TBPacket* packet;
-            var status = tb_client_acquire_packet(client, &packet);
-            switch (status)
+            if (client != IntPtr.Zero)
             {
-                case PacketAcquireStatus.Ok:
-                    AssertTrue(packet != null);
-                    return new Packet(packet);
-                case PacketAcquireStatus.ConcurrencyMaxExceeded:
-                    throw new ConcurrencyExceededException();
-                case PacketAcquireStatus.Shutdown:
-                    throw new ObjectDisposedException("Client is closing");
-                default:
-                    throw new NotImplementedException();
+                lock (this)
+                {
+                    if (client != IntPtr.Zero)
+                    {
+                        tb_client_submit(client, packet);
+                        return;
+                    }
+                }
```

Comment:
Replace the `lock` with the same pattern for synchronization used by the other clients.
In C# we can use `Interlocked` for atomic operations.

Reply:
Thoughts on using a mutex for Go and `synchronized` for Java instead (replacing atomics)?

Reply:
Usually we would avoid locking a mutex in the hot-path like `submit`, but I'm not sure if it would be called with a high enough frequency to be a noticeable difference.

Anyway, in `packet.zig` we already have a concurrent list guarded by a mutex, maybe we could get rid of this one and synchronize from the language side only ...

Reply:
Never mind, we don't need a `ConcurrentStack` anymore. It only held free packets (I thought it held queued requests instead).

So, please drop the type in `packet.zig` then! :tada:

Reply:
> Think it then comes down to a question of simplicity vs practical scalability. Submitting packets likely wont be the bottleneck of a client for some time.

Agreed, no premature optimizations! :+1: